go/mounter/keybase-mount-helper/main_test.go

```diff
@@ -0,0 +1,127 @@
+// license that can be found in the LICENSE file.
+
+// +build !windows
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestFirstMount(t *testing.T) {
+	tempdir, err := ioutil.TempDir(os.TempDir(), "keybasemounthelper")
+	require.NoError(t, err)
+	defer func() {
+		err := os.RemoveAll(tempdir)
+		assert.NoError(t, err)
+	}()
+
+	originalTarget := filepath.Join(tempdir, "target")
+	err = os.Mkdir(originalTarget, 0700)
+	require.NoError(t, err)
+
+	firstMountLink := filepath.Join(tempdir, "mount1")
+	err = os.Symlink(originalTarget, firstMountLink)
+	require.NoError(t, err)
+
+	userMount1 := filepath.Join(tempdir, "user1")
+	err = os.Mkdir(userMount1, 0700)
+	require.NoError(t, err)
+
+	t.Log("First attempt should succeed and create the link")
+	firstMount, err := checkAndSwitchMount(
+		firstMountLink, originalTarget, userMount1)
+	require.NoError(t, err)
+	require.True(t, firstMount)
+	currLink, err := os.Readlink(firstMountLink)
+	require.NoError(t, err)
+	require.Equal(t, userMount1, currLink)
+
+	t.Log("Second attempt shouldn't succeed; no link created")
+	userMount2 := filepath.Join(tempdir, "user2")
+	err = os.Mkdir(userMount2, 0700)
+	require.NoError(t, err)
+	firstMount, err = checkAndSwitchMount(
+		firstMountLink, originalTarget, userMount2)
+	require.NoError(t, err)
+	require.False(t, firstMount)
+	currLink, err = os.Readlink(firstMountLink)
+	require.NoError(t, err)
+	require.Equal(t, userMount1, currLink)
+}
+
+func TestFirstMountRace(t *testing.T) {
+	tempdir, err := ioutil.TempDir(os.TempDir(), "keybasemounthelper")
+	require.NoError(t, err)
+	defer func() {
+		err := os.RemoveAll(tempdir)
+		assert.NoError(t, err)
+	}()
+
+	originalTarget := filepath.Join(tempdir, "target")
+	err = os.Mkdir(originalTarget, 0700)
+	require.NoError(t, err)
+
+	firstMountLink := filepath.Join(tempdir, "mount1")
+	err = os.Symlink(originalTarget, firstMountLink)
+	require.NoError(t, err)
+
+	userMount1 := filepath.Join(tempdir, "user1")
+	err = os.Mkdir(userMount1, 0700)
+	require.NoError(t, err)
+
+	userMount2 := filepath.Join(tempdir, "user2")
+	err = os.Mkdir(userMount2, 0700)
+	require.NoError(t, err)
+
+	fileLock, err := takeLock(firstMountLink)
+	require.NoError(t, err)
+	defer fileLock.Close()
+
+	t.Log("First user with lock takes the symlink in a goroutine")
+	errCh1 := make(chan error, 1)
+	go func() {
+		err = os.Remove(firstMountLink)
+		if err != nil {
+			errCh1 <- err
+			return
+		}
+		errCh1 <- os.Symlink(userMount1, firstMountLink)
+	}()
+
+	t.Log("Second user without lock has to block, " +
+		"then will fail to get the mount")
+	firstMountCh := make(chan bool, 1)
+	errCh2 := make(chan error, 1)
+	go func() {
+		firstMount, err := checkAndSwitchMount(
+			firstMountLink, originalTarget, userMount2)
+		firstMountCh <- firstMount
+		errCh2 <- err
+	}()
+
+	err = <-errCh1
+	require.NoError(t, err)
```

Comment:
Perhaps add a `select` here on `firstMountCh` and a `1s` timeout to make sure 2nd user indeed blocked?

Reply:
I didn't want to add any forced latency into the test.

Reply:
Hmm but this is kind of critical right? How about 200ms 😜

Reply:
I don't agree that it's critical to test this, and in any case if the second user does not properly get blocked, then some runs of the test will fail later because `firstMount` will be `true`.  That's the whole point of the test.

Reply:
Oh that makes sense. Sounds good then. Thanks for the explanation!